File: services/flow-engine/variable-resolver.ts
/**
 * VariableResolver — Resolução de {{variáveis}} em textos
 *
 * Substitui placeholders `{{nomeVariável}}` pelo valor correspondente
 * no contexto de execução do flow (variáveis da sessão + contato + sistema).
 *
 * @see docs/interative_message_flow_builder.md §14.4
 */

import log from "@/lib/log";
import type { DeliveryContext } from "@/types/flow-engine";

// Regex: captura conteúdo entre {{ e }}, ignorando espaços internos
const VARIABLE_PATTERN = /\{\{\s*([a-zA-Z0-9_.]+)\s*\}\}/g;

export class VariableResolver {
	private readonly context: DeliveryContext;
	private readonly sessionVars: Record<string, unknown>;

	constructor(context: DeliveryContext, sessionVariables: Record<string, unknown> = {}) {
		this.context = context;
		this.sessionVars = sessionVariables;
	}

	// ---------------------------------------------------------------------------
	// Public API
	// ---------------------------------------------------------------------------

	/**
	 * Resolve todas as `{{variáveis}}` em uma string.
	 * Variáveis não encontradas ficam inalteradas.
	 */
	resolve(template: string): string {
		if (!template || !template.includes("{{")) return template;

		return template.replace(VARIABLE_PATTERN, (_match, varName: string) => {
			const value = this.lookup(varName);
			if (value === undefined) {
				log.debug("[VariableResolver] Variável não encontrada, mantendo placeholder", { varName });
				return `{{${varName}}}`;
			}
			return String(value);
		});
	}

	/**
	 * Resolve variáveis em todos os campos string de um objeto (shallow).
	 */
	resolveObject<T extends Record<string, unknown>>(obj: T): T {
		const resolved = { ...obj };
		for (const [key, value] of Object.entries(resolved)) {
			if (typeof value === "string") {
				(resolved as Record<string, unknown>)[key] = this.resolve(value);
			}
		}
		return resolved;
	}

	/**
	 * Retorna todas as variáveis disponíveis (para UI de highlight / autocomplete).
	 */
	getAvailableVariables(): Array<{ name: string; value: string; source: string }> {
		const vars: Array<{ name: string; value: string; source: string }> = [];

		// Sistema
		vars.push(
			{ name: "system_timestamp", value: new Date().toISOString(), source: "system" },
			{ name: "system_date", value: new Date().toLocaleDateString("pt-BR"), source: "system" },
			{ name: "system_time", value: new Date().toLocaleTimeString("pt-BR"), source: "system" },
		);

		// Contato
		vars.push(
			{ name: "contact_name", value: this.context.contactName ?? "", source: "contact" },
			{ name: "contact_phone", value: this.context.contactPhone ?? "", source: "contact" },
			{ name: "contact_id", value: String(this.context.contactId), source: "contact" },
		);

		// Conversa
		vars.push(
			{ name: "conversation_id", value: String(this.context.conversationId), source: "conversation" },
			{ name: "conversation_channel", value: this.context.channelType ?? "", source: "conversation" },
			{ name: "conversation_inbox_id", value: String(this.context.inboxId), source: "conversation" },
		);

		// Variáveis da sessão
		for (const [key, value] of Object.entries(this.sessionVars)) {
			vars.push({
				name: key,
				value: value !== undefined && value !== null ? String(value) : "",
				source: "session",
			});
		}

		return vars;
	}

	/**
	 * Atualiza variável da sessão (usado por SET_VARIABLE node).
	 */
	setVariable(name: string, value: unknown): void {
		this.sessionVars[name] = value;
	}

	getVariable(name: string): unknown {
		return this.sessionVars[name];
	}

	/**
	 * Retorna uma cópia das variáveis da sessão (para persistir no Prisma).
	 */
	getSessionVariables(): Record<string, unknown> {
		return { ...this.sessionVars };
	}

	// ---------------------------------------------------------------------------
	// Lookup chain
	// ---------------------------------------------------------------------------

	private lookup(varName: string): unknown {
		// 1. Variáveis de sessão (maior prioridade)
		if (varName in this.sessionVars) {
			return this.sessionVars[varName];
		}

		// Normalizar: aceitar tanto dot notation (contact.name) quanto underscore (contact_name)
		const normalized = varName.includes("_") && !varName.includes(".")
			? varName.replace(/^(contact|conversation|system)_/, "$1.")
			: varName;

		// 2. Variáveis de contato (prefixo contact.)
		if (normalized.startsWith("contact.")) {
			return this.lookupContact(normalized.slice("contact.".length));
		}

		// 3. Variáveis de conversa (prefixo conversation.)
		if (normalized.startsWith("conversation.")) {
			return this.lookupConversation(normalized.slice("conversation.".length));
		}

		// 4. Variáveis de sistema (prefixo system.)
		if (normalized.startsWith("system.")) {
			return this.lookupSystem(normalized.slice("system.".length));
		}

		// 5. Fallback: buscar sem prefixo nas variáveis de sessão com dot notation
		return this.lookupNested(this.sessionVars, varName);
	}

	private lookupContact(field: string): unknown {
		const contactMap: Record<string, unknown> = {
			name: this.context.contactName,
			phone: this.context.contactPhone,
			id: this.context.contactId,
		};
		return contactMap[field];
	}

	private lookupConversation(field: string): unknown {
		const convMap: Record<string, unknown> = {
			id: this.context.conversationId,
			channel: this.context.channelType,
			inbox_id: this.context.inboxId,
			account_id: this.context.accountId,
		};
		return convMap[field];
	}

	private lookupSystem(field: string): unknown {
		const now = new Date();
		const systemMap: Record<string, unknown> = {
			timestamp: now.toISOString(),
			date: now.toLocaleDateString("pt-BR"),
			time: now.toLocaleTimeString("pt-BR"),
			epoch: now.getTime(),
		};
		return systemMap[field];
	}

	/**
	 * Resolve dot notation em objeto aninhado.
	 * Ex: `lookupNested({user: {name: "João"}}, "user.name")` → "João"
	 */
	private lookupNested(obj: Record<string, unknown>, path: string): unknown {
		const parts = path.split(".");
		let current: unknown = obj;

		for (const part of parts) {
			if (current === null || current === undefined || typeof current !== "object") {
				return undefined;
			}
			current = (current as Record<string, unknown>)[part];
		}

		return current;
	}
}

File: domains/socialwise/services/flow/variable_resolver.py
"""VariableResolver — Resolve ``{{variables}}`` in text templates.

Port of services/flow-engine/variable-resolver.ts.

Substitutes placeholders ``{{varName}}`` with the corresponding value
from the flow execution context (session vars + contact + system).
"""

from __future__ import annotations

import re
from datetime import datetime
from typing import Any

from platform_core.logging.config import get_logger

logger = get_logger(__name__)

VARIABLE_PATTERN = re.compile(r"\{\{\s*([a-zA-Z0-9_.]+)\s*\}\}")

# Brazil timezone for system date/time
try:
    from zoneinfo import ZoneInfo
    _BR_TZ = ZoneInfo("America/Sao_Paulo")
except Exception:
    _BR_TZ = None


class VariableResolver:
    """Resolve ``{{var}}`` placeholders in text strings."""

    def __init__(
        self,
        context: Any,
        session_variables: dict[str, Any] | None = None,
    ) -> None:
        self._ctx = context
        self._session_vars: dict[str, Any] = dict(session_variables or {})

    # -------------------------------------------------------------------
    # Public API
    # -------------------------------------------------------------------

    def resolve(self, template: str) -> str:
        """Resolve all ``{{variables}}`` in a string. Unknown vars stay unchanged."""
        if not template or "{{" not in template:
            return template

        def _replacer(match: re.Match) -> str:
            var_name = match.group(1)
            value = self._lookup(var_name)
            if value is None:
                return f"{{{{{var_name}}}}}"
            return str(value)

        return VARIABLE_PATTERN.sub(_replacer, template)

    def resolve_object(self, obj: dict[str, Any]) -> dict[str, Any]:
        """Resolve variables in all string values of a dict (shallow)."""
        resolved = dict(obj)
        for key, value in resolved.items():
            if isinstance(value, str):
                resolved[key] = self.resolve(value)
        return resolved

    def set_variable(self, name: str, value: Any) -> None:
        self._session_vars[name] = value

    def get_variable(self, name: str) -> Any:
        return self._session_vars.get(name)

    def get_session_variables(self) -> dict[str, Any]:
        return dict(self._session_vars)

    def get_available_variables(self) -> list[dict[str, str]]:
        """Return all available variables (for UI autocomplete)."""
        now = _now_br()
        variables: list[dict[str, str]] = [
            {"name": "system_timestamp", "value": now.isoformat(), "source": "system"},
            {"name": "system_date", "value": now.strftime("%d/%m/%Y"), "source": "system"},
            {"name": "system_time", "value": now.strftime("%H:%M:%S"), "source": "system"},
            {"name": "contact_name", "value": str(self._ctx_get("contact_name", "")), "source": "contact"},
            {"name": "contact_phone", "value": str(self._ctx_get("contact_phone", "")), "source": "contact"},
            {"name": "contact_id", "value": str(self._ctx_get("contact_id", "")), "source": "contact"},
            {"name": "conversation_id", "value": str(self._ctx_get("conversation_id", "")), "source": "conversation"},
            {"name": "conversation_channel", "value": str(self._ctx_get("channel_type", "")), "source": "conversation"},
            {"name": "conversation_inbox_id", "value": str(self._ctx_get("inbox_id", "")), "source": "conversation"},
        ]
        for key, value in self._session_vars.items():
            variables.append({
                "name": key,
                "value": str(value) if value is not None else "",
                "source": "session",
            })
        return variables

    # -------------------------------------------------------------------
    # Lookup chain
    # -------------------------------------------------------------------

    def _lookup(self, var_name: str) -> Any:
        # 1. Session variables (highest priority)
        if var_name in self._session_vars:
            return self._session_vars[var_name]

        # Normalize: accept both dot notation and underscore
        normalized = var_name
        if "_" in var_name and "." not in var_name:
            for prefix in ("contact", "conversation", "system"):
                if var_name.startswith(f"{prefix}_"):
                    normalized = var_name.replace(f"{prefix}_", f"{prefix}.", 1)
                    break

        # 2. Contact variables
        if normalized.startswith("contact."):
            return self._lookup_contact(normalized[len("contact."):])

        # 3. Conversation variables
        if normalized.startswith("conversation."):
            return self._lookup_conversation(normalized[len("conversation."):])

        # 4. System variables
        if normalized.startswith("system."):
            return self._lookup_system(normalized[len("system."):])

        # 5. Fallback: nested dot notation in session vars
        return self._lookup_nested(self._session_vars, var_name)

    def _lookup_contact(self, field: str) -> Any:
        contact_map = {
            "name": self._ctx_get("contact_name"),
            "phone": self._ctx_get("contact_phone"),
            "id": self._ctx_get("contact_id"),
        }
        return contact_map.get(field)

    def _lookup_conversation(self, field: str) -> Any:
        conv_map = {
            "id": self._ctx_get("conversation_id"),
            "channel": self._ctx_get("channel_type"),
            "inbox_id": self._ctx_get("inbox_id"),
            "account_id": self._ctx_get("account_id"),
        }
        return conv_map.get(field)

    def _lookup_system(self, field: str) -> Any:
        now = _now_br()
        system_map: dict[str, Any] = {
            "timestamp": now.isoformat(),
            "date": now.strftime("%d/%m/%Y"),
            "time": now.strftime("%H:%M:%S"),
            "epoch": int(now.timestamp() * 1000),
        }
        return system_map.get(field)

    @staticmethod
    def _lookup_nested(obj: dict[str, Any], path: str) -> Any:
        """Resolve dot notation in nested dicts."""
        parts = path.split(".")
        current: Any = obj
        for part in parts:
            if current is None or not isinstance(current, dict):
                return None
            current = current.get(part)
        return current

    def _ctx_get(self, key: str, default: Any = None) -> Any:
        if isinstance(self._ctx, dict):
            return self._ctx.get(key, default)
        if hasattr(self._ctx, key):
            return getattr(self._ctx, key, default)
        return default


def _now_br() -> datetime:
    """Return current time in Brazil timezone."""
    if _BR_TZ:
        return datetime.now(_BR_TZ)
    return datetime.utcnow()
